src/frequenz/channels/_receiver.py

```diff
@@ -336,9 +355,93 @@ def consume(self) -> MappedMessageT_co:  # noqa: DOC502
 
     def __repr__(self) -> str:
-        """Return a string representation of the timer."""
+        """Return a string representation of the mapper."""
         return f"{type(self).__name__}({self._receiver!r}, {self._mapping_function!r})"
+
+
+class _Sentinel:
+    """A sentinel object to represent no value received yet."""
+
+    def __str__(self) -> str:
+        """Return a string representation of this sentinel."""
+        return "<No message ready to be consumed>"
+
+
+_SENTINEL = _Sentinel()
+
+
+class _Filter(Receiver[ReceiverMessageT_co], Generic[ReceiverMessageT_co]):
+    """Apply a filter function on the messages on a receiver."""
+
+    def __init__(
+        self,
+        *,
+        receiver: Receiver[ReceiverMessageT_co],
+        filter_function: Callable[[ReceiverMessageT_co], bool],
+    ) -> None:
+        """Initialize this receiver filter.
+
+        Args:
+            receiver: The input receiver.
+            filter_function: The function to apply on the input data.
+        """
+        self._receiver: Receiver[ReceiverMessageT_co] = receiver
+        """The input receiver."""
+
+        self._filter_function: Callable[[ReceiverMessageT_co], bool] = filter_function
+        """The function to apply on the input data."""
+
+        self._next_message: ReceiverMessageT_co | _Sentinel = _SENTINEL
+
+        self._recv_closed = False
+
+    async def ready(self) -> bool:
+        """Wait until the receiver is ready with a message or an error.
+
+        Once a call to `ready()` has finished, the message should be read with
+        a call to `consume()` (`receive()` or iterated over). The receiver will
+        remain ready (this method will return immediately) until it is
+        consumed.
+
+        Returns:
+            Whether the receiver is still active.
+        """
+        while await self._receiver.ready():
+            message = self._receiver.consume()
+            if self._filter_function(message):
+                self._next_message = message
+                return True
+        self._recv_closed = True
+        return False
+
+    def consume(self) -> ReceiverMessageT_co:
+        """Return a transformed message once `ready()` is complete.
+
+        Returns:
+            The next message that was received.
+
+        Raises:
+            ReceiverStoppedError: If the receiver stopped producing messages.
+            ReceiverError: If there is a problem with the receiver.
+        """
+        if self._recv_closed:
+            raise ReceiverStoppedError(self)
+        assert not isinstance(
+            self._next_message, _Sentinel
+        ), "`consume()` must be preceded by a call to `ready()`"
+
+        message = self._next_message
+        self._next_message = _SENTINEL
+        return message
+
+    def __str__(self) -> str:
+        """Return a string representation of the filter."""
+        return f"{type(self).__name__}:{self._receiver}:{self._filter_function}"
```

Comment:
Wouldn't it be useful here to print also the `_next_message`?

Reply:
I think it doesn't make sense, because it would only have a value between calls to `ready` and `consume`.  At all other times, it would only print the sentinel.  So I don't think there's any value in it.

Reply:
Right, is not like `LatestValueCache`.

Reply:
But then if the `Sentinel` class is never printed nor exposed outside, there is not much point in having either of `__str__` nor `__repr__`, unless we use it for internal debugging, in which case I think it makes more sense to have `__repr__` as normally pytest and debuggers will use that one.

I think I would print the `_next_value` in this classes `__repr__`, as when debugging or in a pytest it might be useful to see what are its contents, even if most of the time is the sentinel. When calling `repr()` you usually want to know the internal state of the instance. Usually `<` `>` is used to represent internal state when `repr()` doesn't match the code used to create the class, like the default `__repr__` in object and what's used to print functions or methods:

```py
>>> class C: pass
... 
>>> repr(C())
'<__main__.C object at 0x7f564f233690>'
>>> repr(C.__init__)
"<slot wrapper '__init__' of 'object' objects>"
```

Reply:
done